packages/react/popper/src/Popper.tsx

```diff
@@ -18,17 +18,78 @@ import type { Measurable } from '@radix-ui/rect';
 const POPPER_NAME = 'Popper';
 
 type PopperContextValue = {
+  anchorRef: React.MutableRefObject<Measurable | null>;
+};
+const [PopperProvider, usePopperContext] = createContext<PopperContextValue>(POPPER_NAME);
+
+const Popper: React.FC = ({ children }) => {
+  const anchorRef = React.useRef<Measurable | null>(null);
+  return <PopperProvider anchorRef={anchorRef}>{children}</PopperProvider>;
+};
+
+Popper.displayName = POPPER_NAME;
+
+/* -------------------------------------------------------------------------------------------------
+ * PopperAnchor
+ * -----------------------------------------------------------------------------------------------*/
+
+const ANCHOR_NAME = 'PopperAnchor';
+
+type PopperAnchorOwnProps = Polymorphic.Merge<
+  Polymorphic.OwnProps<typeof Primitive>,
+  { virtualRef?: React.RefObject<Measurable> }
+>;
+type PopperAnchorPrimitive = Polymorphic.ForwardRefComponent<
+  Polymorphic.IntrinsicElement<typeof Primitive>,
+  PopperAnchorOwnProps
+>;
+
+const PopperAnchor = React.forwardRef((props, forwardedRef) => {
+  const { virtualRef, children, ...anchorProps } = props;
+  const context = usePopperContext(ANCHOR_NAME);
+  const ref = React.useRef<React.ElementRef<typeof Primitive>>(null);
+  const composedRefs = useComposedRefs(
+    forwardedRef,
+    ref,
+    context.anchorRef as React.MutableRefObject<React.ElementRef<typeof Primitive>>
+  );
+
+  React.useEffect(() => {
+    if (virtualRef?.current) {
+      context.anchorRef.current = virtualRef.current;
+    }
+  });
```

Comment:
Might be a good idea to add a comment here as to why we do this.

Reply:
```
// Consumer can anchor the popper to something that isn't
// a DOM node e.g. pointer position, so we override the
// `anchorRef` with their virtual ref in this case.
```

Is this okay? I'll add it for now and we can tweak

Reply:
Looks good to me.